File: medium/17-letter-combination-phone-number.cpp
class Solution {
public:
  vector<string> hash {"abc","def","ghi","jkl","mno","pqrs","tuv","wxyz"};
  void dp(int i, string dig, string curr, vector<string>& ans)
  {
    if(i == dig.size())
      return ans.push_back(curr);
    int digit = dig[i]-'0';
    for(int k = 0; k < hash[digit-2].size(); k++)
      dp(i+1, dig, curr + hash[digit-2][k], ans);
  }
  vector<string> letterCombinations(string digits) {
    vector<string> ans;
    if(digits.size() == 0) return ans;
    dp(0, digits, "", ans);
    return ans;
  }
};

File: medium/17-letter-combination-phone-number.js
/**
 * @param {string} digits
 * @return {string[]}
 */
 var letterCombinations = function(digits) {
  
  if (!digits.length) return [];
  
  const map = {
    1: "",
    2: "abc",
    3: "def",
    4: "ghi",
    5: "jkl",
    6: "mno",
    7: "pqrs",
    8: "tuv",
    9: "wxyz"
  };
  
  function solve(i, digits, cur, answer) {
    if (i == digits.length) {
      return answer.push(cur);
    }
    const digit = parseInt(digits[i], 10);
    
    for (let k = 0; k < map[digit].length; k++) {
      solve(i+1, digits, cur + map[digit][k], answer);
    }
  }
  
  const answer = [];
  solve(0, digits, "", answer);
  return answer;
  
};